File: src/main/java/io/aif/associations/builder/ExperimentsConnectionsGraphReducer.java
package io.aif.associations.builder;


import java.util.*;
import java.util.stream.Collectors;

class ExperimentsConnectionsGraphReducer<T> {
    
    public Map<T, Map<T, Double>> reduce(final Map<T, Map<T, List<Double>>> originalGraph) {
        final Map<T, Map<T, Double>> resultsGraph = convertGraph(originalGraph);

        normalize(resultsGraph);
        
        return resultsGraph;
    }
    
    private void normalize(final Map<T, Map<T, Double>> graph) {
        if (graph.isEmpty()) return;

        final Double maxEdge = graph.entrySet().stream().flatMap(entry -> entry.getValue().entrySet().stream()).mapToDouble(Map.Entry::getValue).max().getAsDouble();
        final Double minEdge = graph.entrySet().stream().flatMap(entry -> entry.getValue().entrySet().stream()).mapToDouble(Map.Entry::getValue).min().getAsDouble();
        final Double newMax = maxEdge - minEdge > 0 ? maxEdge - minEdge : maxEdge;
        graph.keySet().forEach(from ->
            graph.get(from).keySet().forEach(to -> {
                        final Double weight = graph.get(from).get(to);
                        graph.get(from).put(to, ((newMax - (weight - minEdge)) / newMax));
                    }
            )
        );
    }
    
    private Map<T, Map<T, Double>> convertGraph(final Map<T, Map<T, List<Double>>> originalGraph) {
        return originalGraph.entrySet().stream().map(globalEntry ->
                        new HashMap.SimpleEntry<T, Map<T, Double>>(globalEntry.getKey(),
                                convertVertexEdges(globalEntry.getValue())
                        )
        ).collect(Collectors.toMap(
                Map.Entry::getKey,
                Map.Entry::getValue,
                (v1, v2) -> v1
        ));
    }

    private Map<T, Double> convertVertexEdges(final Map<T, List<Double>> vertexEdges) {
        return vertexEdges.entrySet().stream()
                .map(entry ->
                    new HashMap.SimpleEntry<>(entry.getKey(), entry.getValue().stream().mapToDouble(x -> x).sum()))
                .collect(Collectors.toMap(
                    Map.Entry::getKey,
                    Map.Entry::getValue,
                    (v1, v2) -> v1
                ));
    }
    
}

File: src/main/scala/io/aif/associations/builder/ExperimentsConnectionsGraphReducer.scala
package io.aif.associations.builder

import io.aif.associations.model.graph.Graph


class ExperimentsConnectionsGraphReducer[T] {
  
  def reduce(originalGraph: Graph[T, List[Double]]): Graph[T, Double] = {
    
    def convertVertexConnections(vertexConnections: Map[T, List[Double]]) = {
      vertexConnections.keys.map(key => (key, vertexConnections.get(key).get.sum )) toMap
    }
    
    def max(graph: Map[T, Map[T, Double]]) = {
      graph.values.map(connection => connection.values max) max
    }

    def min(graph: Map[T, Map[T, Double]]) = {
      graph.values.map(connection => connection.values min) min
    }

    def normalize(graph: Map[T, Map[T, Double]], max : Double, min : Double) = {
      val newMax = if (max - min > 0) max - min else max
      graph mapValues(vertexConnections => vertexConnections.mapValues(weight => ((newMax - (weight - min)) / newMax)))
    }

    val rawData = originalGraph.rawData

    rawData match {
      case rawData.isEmpty => new Graph[T, Double](Map())
      case _ => {
        val recalculateGraph = rawData.mapValues(convertVertexConnections)
        new Graph[T, Double](normalize(recalculateGraph, max(recalculateGraph), min(recalculateGraph)))
      }
    }
  }

}
